### ISSUE TITLE
cmd/govim: add support for quickfix item signs

### ISSUE THREAD

**Issue by leitzler** [myitcv/govim#284]
This is a proposal.

It would be nice if quickfix items could be annotated using vim's built in "sign".
There is a rough proof-of-concept implementation at https://github.com/leitzler/govim (master) that defines a new sign named "err" (ought to be something with "govim") using:
```
sign define err text=>> texthl=Error
```
It tells vim to place ">>" using the "Error" highlight when the sign named "err" is used.

And then each time the quickfix list is populated it removes all signs (regardless of buffer or file, can be improved) using:
```
sign unplaced *
```

.. to finally place a new sign next to each item:
```
sign place <Id> line=<lineno> name=err file=<filename>
```

Note that due to gopls dropping some diagnostics the sign placing won't be as consistent as it should.

**Comment by myitcv** [comment]
This sounds very interesting, thank you; can you share some screenshots/gifs here to help others visualise this quickly?

**Comment by myitcv** [comment]
Just had a quick play with your POC. Looks great! Thanks for putting together a straw man; makes it so much easier to discuss/iterate.

> And then each time the quickfix list is populated it removes all signs (regardless of buffer or file, can be improved

There's quite a lot of flickering with this approach in my Vim setup. 

To start with I had `signcolumn=auto`; so the "gutter" itself was disappearing then reappearing with each key stroke. This is obviously easily fixed. 

But also the red markers themselves started to look a bit like a disco!

I suspect the only way to do this will be to iterate through the new list of diagnostics for a file _and_ the list of current signs and work out when/where to add/remove signs?

**Comment by leitzler** [comment]
I *think* a lot of the flickering is due to https://github.com/golang/go/issues/32443#issuecomment-499113570 - at least my quickfix flickers as much as the signs.

In my work setup where I use ale it actually places signs as I type, and it isn’t annoying at all. Need to play around a bit more with it.

**Comment by myitcv** [comment]
> I think a lot of the flickering is due to golang/go#32443 (comment) - at least my quickfix flickers as much as the signs.

Yes https://github.com/golang/go/issues/32443#issuecomment-499113570 will create flicker, but if you try your POC with a single file-package, then you will also see the flicker. I suspect that's because a redraw happens after you clear the signs. Hence why with signs I think you need to combine the two states to reduce the flicker (this shouldn't be necessary with the quickfix because we send the entire new contents of the quickfix each time and Vim works out the redraw delta)

**Comment by leitzler** [comment]
I did some work yesterday on the poc (not yet pushed) where I, instead of keeping track of placed signs inside govim, ask vim for all placed signs. That will ensure a consistent state (as I noticed that signs might move as the buffer changes).

I also added a sign group to scope govim signs, to avoid collisions with other plugins that uses signs.

When using sign groups it isn’t however possible to get all signs across all buffers, so I will have to query each of the open buffers. As soon as I have that in place I’ll push the changes.

Oh, and I use functions for all sign operations.

It looks quite good so far anyway!

**Comment by leitzler** [comment]
I've updated my fork with the latest changes. Haven't had time to test it yet but it should at least show the idea now.

I did notice that when I open a file with lots of quickfix entries (20+) the sign's are placed in chunks with some latency. I guess it could have to do with the fact that there is loads of calls to vim since they can't be batch-placed.

**Comment by myitcv** [comment]
Per Slack chat:

* just to note - that there will be a delay in signs _and_ the quickfix window updating following #294 
* if we need batch support for signs, then we can raise a Vim issue

**Comment by myitcv** [comment]
Raised https://groups.google.com/forum/#!topic/vim_dev/aW2buOgZ6qI to try and get some help/insight on the stalls you're seeing.

**Comment by myitcv** [comment]
Per discussions with @leitzler, this may or may not be dependent on https://github.com/vim/vim/issues/4557

**Comment by leitzler** [comment]
Vim version 1.8.1605 fixes the staling behaviour and I have rebased my fork since.

It seems to work but needs more testing, and I think it might be a good idea to wait for https://github.com/vim/vim/issues/4557 anyway since it will improve UX a bunch. Vim has a delay of 10ms between channel reads in certain cases and that will introduce a latency if there are many signs.

**Comment by myitcv** [comment]
Based on some of the discussion in https://github.com/vim/vim/issues/4557, I think a good way to go here (in the short term at least) will be to batch calls from the `govim` side. 

I've pushed up https://github.com/myitcv/govim/pull/341 as a proof of concept in that regard. Feedback welcome on that PR/the approach.

I've also pushed up #342 as an example of how that would be used in practice. Based on the positive experience with #342 at least I'm tempted to merge this because it's a more generic approach that the previous approach to applying format changes.

FYI @milesbxf regarding #341
